### PR TITLE
ARROW-2102: [C++] Implement take kernel functions - primitive value type

### PR DESCRIPTION
**Pull request by alphalfalfa**
Currently the take kernel supports:
- Primitive value types (NA, Boolean and numeric types)
- Single numeric index
- Multiple flat indices in an array

The kernel does not yet handle special indices including negative integers or out-of-bound indices. ARROW-2105 is created for such feature. Parent ticket is at Arrow-772.

### CODE REVIEW

**Comment by alphalfalfa** [comment]
@wesm @xuepanchen

**Comment by wesm** [comment]
@alphalfalfa thank you for starting on these kernels; I'm sorry it's taken me a little while to get down to review but I intend to do so this week

File: cpp/src/arrow/compute/kernels/take.cc [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+template <typename I>
+struct TakeFunctor<BooleanType, I, enable_if_index_valid<I>> {
+  void operator()(FunctionContext* ctx, const TakeOptions& options,
+                  const ArrayData& input, const ArrayData& indices, ArrayData* output) {
+    auto index_data = GetValues<int32_t>(indices, 1);
+    const uint8_t* input_data = input.buffers[1]->data();
+    uint8_t* out_data = output->buffers[1]->mutable_data();
+    internal::BitmapWriter bit_writer(out_data, output->offset, indices.length);
+    for (int64_t i = 0; i < indices.length; ++i) {
+      if (BitUtil::GetBit(input_data, *index_data++)) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/take.cc]: Ditto with null bitmap.

File: cpp/src/arrow/compute/kernels/take.cc [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    memset(validity_bitmap->mutable_data(), 0, bitmap_size);
+  } else {
+    out->null_count = 0;
+    memset(validity_bitmap->mutable_data(), 0xFF, bitmap_size); // preset all to valid
+  }
+
+  const Type::type type_id = out->type->id();
+  if (!(is_primitive(type_id) || type_id == Type::FIXED_SIZE_BINARY ||
+      type_id == Type::DECIMAL)) {
+    std::stringstream ss;
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/take.cc]: No need stringstream, see other comment on the subject.

File: cpp/src/arrow/compute/kernels/take.cc [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+template <typename I>
+struct TakeFunctor<NullType, I, enable_if_index_valid<I>> {
+  void operator()(FunctionContext* ctx, const TakeOptions& options,
+                  const ArrayData& input, const ArrayData& indices, ArrayData* output) {}
+};
+
+// ----------------------------------------------------------------------
+
+typedef std::function<void(FunctionContext*,
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/take.cc]: Why the typedef, can it just pass the Functor arround?

File: cpp/src/arrow/compute/kernels/take.cc [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+  out->buffers.push_back(out_data);
+
+  return Status::OK();
+}
+
+template <typename IndexType>
+class TakeKernel : public UnaryKernel {};
+
+template <>
+class TakeKernel<Int32Type> : public UnaryKernel {
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/take.cc]: Why `Int32Type`, this seems fishy as it doesn't seem used in the rest of the code.

File: cpp/src/arrow/compute/kernels/take.cc [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+  FN(VALUE_TYPE, UInt8Type);          \
+  FN(VALUE_TYPE, Int16Type);          \
+  FN(VALUE_TYPE, UInt16Type);         \
+  FN(VALUE_TYPE, Int32Type);          \
+  FN(VALUE_TYPE, UInt32Type);         \
+  FN(VALUE_TYPE, Int64Type);          \
+  FN(VALUE_TYPE, UInt64Type);
+
+#define GET_TAKE_FUNCTION(CASE_GENERATOR, ValueType)                           \
+  static std::unique_ptr<UnaryKernel> Get##ValueType##TakeFunc(                \
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/take.cc]: I'd make this function return Status and pass the unique_ptr by output parameter (as the parent callsite is also doing).

File: cpp/src/arrow/compute/kernels/take.cc [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+  TakeFunction func_;
+  std::shared_ptr<DataType> index_type_;
+};
+
+#define TAKE_CASE(ValueType, IndexType)                                                 \
+  case IndexType::type_id:                                                              \
+    func = [](FunctionContext* ctx, const TakeOptions& options, const ArrayData& input, \
+              const ArrayData& indices, ArrayData* out) {                               \
+      TakeFunctor<ValueType, IndexType> func;                                           \
+      func(ctx, options, input, indices, out);                                          \
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/take.cc]: There's way too much `func` shadowing involved here.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

**Comment by wesm** [comment]
As this PR is over a year old I think it probably needs a fresh start

**Comment by wesm** [comment]
Closing in favor of new PR, thank you for starting on this @alphalfalfa!

**Pull request closed without merging**